pkg/users/group-create-dialog.js

```diff
@@ -0,0 +1,169 @@
+            </FormGroup>
+
+            <FormGroup label={_("ID")}
+                       hasNoPaddingTop
+                       helperTextInvalid={errors && errors.id}
+                       validated={(errors && errors.id) ? "error" : "default"}
+                       isStack
+                       fieldId="groups-create-id">
+                <TextInput id="groups-create-id"
+                           validated={(errors && errors.id) ? "error" : "default"}
+                           value={id} onChange={value => change("id", value)} />
+            </FormGroup>
+        </Form>
+    );
+}
+
+function validate_name(name, groups) {
+    if (!name)
+        return _("No group name specified");
+
+    for (let i = 0; i < name.length; i++) {
+        if (!is_valid_char_name(name[i]))
+            return _("The group name can only consist of letters from a-z, digits, dots, dashes and underscores.");
+    }
+
+    for (let k = 0; k < groups.length; k++) {
+        if (groups[k].name == name)
+            return _("A group with this name already exists");
+    }
+
+    return null;
+}
+
+function validate_group(id, groups) {
+    if (!id)
+        return _("No ID specified");
+
+    const id_num = parseInt(id);
+    if (id_num.toString() !== id || id_num < 0)
+        return _("The group ID must be positive integer.");
+
+    for (let k = 0; k < groups.length; k++) {
+        if (groups[k].gid == id_num)
+            return _("A group with this ID already exists");
+    }
+
+    return null;
+}
+
+export function group_create_dialog(groups, setGroupsCardExpanded) {
+    cockpit.file("/etc/login.defs").read()
+            .then(logindef => {
+                const gids = groups
+                        .filter(g => g.name !== 'nobody')
+                        .map(group => group.gid);
+                const min_gid = parseInt(logindef.match(/^GID_MIN\s+(\d+)/m)[1] || 100);
+                const max_gid = parseInt(logindef.match(/^GID_MAX\s+(\d+)/m)[1] || 65534);
+
+                change("id", (Math.max(min_gid, Math.max(...gids.filter(id => id != max_gid)) + 1) + 1).toString());
+            });
+
+    let dlg = null;
+    const state = {
+        name: "",
+        id: "",
+    };
+    let errors = { };
+
+    function change(field, value) {
+        state[field] = value;
+        errors = { };
+
+        update();
+    }
+
+    function validate(name, id) {
+        const errs = { };
+
+        errs.name = validate_name(name, groups);
+        errs.id = validate_group(id, groups);
+        errors = errs;
+
+        return !has_errors(errs);
+    }
+
+    function create(name, id) {
+        const valid = validate(name, id);
+        if (valid) {
+            const group_add_cmd = ["/usr/sbin/groupadd", name, "-g", id];
+
+            return cockpit.spawn(group_add_cmd, { superuser: "require", err: "message" });
+        } else {
+            update();
+            return Promise.reject();
```

Comment:
Looksk a bit weird here to not pass an actual Error (with message) to the caller? I suppose the caller handles validation errors, but it already has to handle failures of cockpit.spawn() as well.

Reply:
We do actually pass the errors, but not with promise rejection but with the errors property in the:

<GroupCreateBody state={state} errors={errors} change={change} /> 

Then we show the errors from validation inline with the form fields, like the are supposed to appear and not on the top of the modal, which generic errors should appear.

Reply:
This is still confusing, though: so you are saying that validation errors are handled in a different way -- but whatever calls this function will have to use .catch() for handling spawn errors; that would also catch this "static" validation error with the empty Promise.reject().